lib/rules/index.ts

```diff
@@ -0,0 +1,29 @@
+import { readdirSync } from 'fs';
+import { join, parse } from 'path';
+
+import { TSESLint } from '@typescript-eslint/experimental-utils';
+
+type RuleModule = TSESLint.RuleModule<string, unknown[]>;
+
+// Copied from https://github.com/babel/babel/blob/b35c78f08dd854b08575fc66ebca323fdbc59dab/packages/babel-helpers/src/helpers.js#L615-L619
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const interopRequireDefault = (obj: any): { default: unknown } =>
+  obj?.__esModule ? obj : { default: obj };
+
+const importDefault = (moduleName: string) =>
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  interopRequireDefault(require(moduleName)).default;
+
+const rulesDir = __dirname;
+const excludedFiles = ['index'];
+
+export default readdirSync(rulesDir)
+  .map((rule) => parse(rule).name)
+  .filter((ruleName) => !excludedFiles.includes(ruleName))
+  .reduce<Record<string, RuleModule>>(
+    (allRules, ruleName) => ({
+      ...allRules,
+      [ruleName]: importDefault(join(rulesDir, ruleName)) as RuleModule,
```

Comment:
Can't we extract here the actual type of each rule? It could be interesting to get that to generate docs automatically for each rule.

Reply:
I can't think of a way to do this tbh, but I'm open to suggestions

Reply:
I guess the type is getting lost by the `importDefault` return value type? Maybe @timdeschryver can help us here 🙏

Reply:
`importDefault` is indeed the issue here I think, but that one needs to be generic in order for my automation to work

Reply:
Would help if we assume we are gonna always have ES module default exports? So we can remove `interopRequireDefault` and simplify `importDefault`, which could help to get the proper type from each module.

Reply:
`importDefault` will then return of type `any`, so that's not the solution you're looking for either I'm afraid 😕

Reply:
Of course... we would have to manually import all rules to get corresponding types. I'll think about it, but we can probably merge this one until we actually need those types.

Reply:
I don't think it's possible to interfer the types here. If needed we could look into adding the types manually since these would all be Rules.

Reply:
The real question is: do we need the types outside of the plugin? 🤔
If not, we're fine as we still have types for our tests, ...

Reply:
We don't need it outside the plugin, that wasn't my idea either. But at some point, we will generate rules docs from rules meta info automatically, and I don't know if at that point we would need the types for being able to generate the docs properly.

Reply:
Docs will never be completely generated automatically, we'll only need meta info for creating the tables in README and for creating configs automatically

That's already included in the `TSESLint.RuleModule` type, so no need to worry about that

Reply:
I wasn't expecting to generate all the docs automatically for sure :) Anyway, it should be enough with `TSESLint.RuleModule` indeed. Thanks!